synapse/lib/ast.py

```diff
@@ -536,19 +549,21 @@ async def todogenr():
 
                 if doedges:
 
-                    async for buid01 in results:
-
+                    async for n2buid, verb in runt.snap.iterEdgePref(node.buid):
                         await asyncio.sleep(0)
+                        if rev.get(n2buid) is None:
+                            rev[n2buid] = collections.defaultdict(list)
 
-                        iden01 = s_common.ehex(buid01)
-                        async for verb in node.iterEdgeVerbs(buid01):
-                            await asyncio.sleep(0)
-                            edges.append((iden01, {'type': 'edge', 'verb': verb}))
+                        # TODO: old way doesn't make self-referential nodes?
+                        if n2buid != node.buid:
+                            rev[n2buid][node.buid].append(verb)
+                            if n2buid in results:
+                                edges.append((s_common.ehex(n2buid), {'type': 'edge', 'verb': verb}))
 
-                        # for existing nodes, we need to add n2 -> n1 edges in reverse
-                        async for verb in runt.snap.iterEdgeVerbs(buid01, node.buid):
+                    for n1buid, verbs in rev.get(node.buid, {}).items():
```

Comment:
If we're fighting for cycles, we may wanna make this get() return a None and check for it rather than construct a new empty dict